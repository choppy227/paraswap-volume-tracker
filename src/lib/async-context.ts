import { NextFunction, Request, Response } from 'express';
import { AsyncLocalStorage } from 'async_hooks';

type AsyncContext = Map<string, any>;

const asyncLocalStorage = new AsyncLocalStorage<AsyncContext>();

export default function AsyncContextMiddleware() {
  return function asyncContextdMiddleware(
    req: Request,
    res: Response,
    next: NextFunction,
  ) {
    asyncLocalStorage.run(new Map(), async () => {
      return next();
    });
  };
}

export function getAsyncContextStore() {
  return asyncLocalStorage.getStore();
}

export function getAsyncContextValue(key: string) {
  return asyncLocalStorage.getStore()?.get(key);
}

export function setAsyncContextValue(key: string, value: any) {
  return asyncLocalStorage.getStore()?.set(key, value);
}
