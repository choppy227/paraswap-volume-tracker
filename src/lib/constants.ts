export const CHAIN_ID_MAINNET = 1;
export const CHAIN_ID_ROPSTEN = 3;
export const CHAIN_ID_BINANCE = 56;
export const CHAIN_ID_POLYGON = 137;
export const CHAIN_ID_AVALANCHE = 43114;

export const STAKING_CHAIN_IDS = [CHAIN_ID_MAINNET, CHAIN_ID_ROPSTEN];
export const STAKING_CHAIN_IDS_SET = new Set([CHAIN_ID_MAINNET, CHAIN_ID_ROPSTEN]);

export const VOLUME_TRACKER_SUPPORTED_NETWORKS = [CHAIN_ID_MAINNET]
export const VOLUME_TRACKER_INIT_TIME: { [network: number]: number } = {
  [CHAIN_ID_MAINNET]: parseInt(process.env.INIT_TIME || '0') //TODO: use the block info to the init time from the init block
}

export const Web3Provider: { [network: number]: string } = {
  [CHAIN_ID_MAINNET]: process.env.HTTP_PROVIDER || '',
  [CHAIN_ID_ROPSTEN]: process.env.HTTP_PROVIDER_3 || '',
  [CHAIN_ID_BINANCE]: process.env.HTTP_PROVIDER_56 || '',
  [CHAIN_ID_POLYGON]: process.env.HTTP_PROVIDER_137 || '',
  [CHAIN_ID_AVALANCHE]: process.env.HTTP_PROVIDER_43114 || '',
};

// TODO: in future we can fetch it from the api directly
export const AugustusV5Address: { [network: number]: string } = {
  [CHAIN_ID_MAINNET]: '0xdef171fe48cf0115b1d80b88dc8eab59176fee57',
  [CHAIN_ID_BINANCE]: '0xdef171fe48cf0115b1d80b88dc8eab59176fee57',
  [CHAIN_ID_POLYGON]: '0xdef171fe48cf0115b1d80b88dc8eab59176fee57',
};

export const AugustusV4Address: { [network: number]: string } = {
  [CHAIN_ID_MAINNET]: '0x1bd435f3c054b6e901b7b108a0ab7617c808677b',
  [CHAIN_ID_BINANCE]: '0x55A0E3b6579972055fAA983482acEb4B251dcF15',
  [CHAIN_ID_POLYGON]: '0x90249ed4d69D70E709fFCd8beE2c5A566f65dADE',
};

export const RewardDistributionAddress: { [network: string]: string } = {
  [CHAIN_ID_MAINNET]: '0x8145cDeeD63e2E3c103F885CbB2cD02a00F54873'
}

// TODO: in future we can fetch it from the api directly
export const ZeroXV2Address: { [network: number]: string } = {
  [CHAIN_ID_MAINNET]: '0x080bf510fcbf18b91105470639e9561022937712',
};

// TODO: in future we can fetch it from the api directly
export const ZeroXV4Address: { [network: number]: string } = {
  [CHAIN_ID_MAINNET]: '0xdef1c0ded9bec7f1a1670819833240f027b25eff',
};

// TODO: set using env variable
export const ParaswapApiURL = 'https://apiv5.paraswap.io';

export const MULTICALL_ADDRESS: any = {
  [CHAIN_ID_MAINNET]: '0xeefba1e63905ef1d7acba5a8513c70307c1ce441',
  [CHAIN_ID_ROPSTEN]: '0x293405FE3aDefDB94A8A1Ed50873a15C6Cc83BC5',
  [CHAIN_ID_BINANCE]: '0xdc6e2b14260f972ad4e5a31c68294fba7e720701',
  [CHAIN_ID_POLYGON]: '0xdC6E2b14260F972ad4e5a31c68294Fba7E720701',
};

export const DEFAULT_CHAIN_ID = parseInt(process.env.DEFAULT_CHAIN_ID || '1');

export const NULL_ADDRESS = '0x0000000000000000000000000000000000000000';

export const NATIVE_TOKEN_ADDRESS =
  '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
