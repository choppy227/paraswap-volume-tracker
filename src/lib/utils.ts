import axios, { AxiosBasicCredentials, AxiosResponse } from 'axios';
import { Request } from 'express';
import * as https from 'https';
import { getRemoteAddress } from './remote-address';

const HTTP_TIMEOUT = 500;

axios.defaults.httpsAgent = new https.Agent({ keepAlive: true });

export class Utils {
  public static readonly isAWS =
    process.env.ECS_CONTAINER_METADATA_URI !== undefined;

  static getIP = (req: Request): string => {
    const address = getRemoteAddress(req);

    return Array.isArray(address)
      ? address[0]
      : address?.split(':')[0] || address;
  };

  static _get<T = any>(
    url: string,
    timeout?: number,
    headers: { [key: string]: string | number } = {},
    auth?: AxiosBasicCredentials,
  ) {
    return Utils._axios<T>(url, 'get', null, timeout, headers, auth);
  }

  static _post<T = any>(
    url: string,
    data?: any,
    timeout?: number,
    headers: { [key: string]: string | number } = {},
    auth?: AxiosBasicCredentials,
  ) {
    return Utils._axios<T>(url, 'post', data, timeout, headers, auth);
  }

  static _axios<T = any>(
    url: string,
    method: 'get' | 'post',
    data?: any,
    timeout: number = HTTP_TIMEOUT,
    headers: { [key: string]: string | number } = {},
    auth?: AxiosBasicCredentials,
  ): Promise<AxiosResponse<T>> {
    return axios({
      method,
      url,
      data,
      timeout,
      headers: {
        'User-Agent': 'node.js',
        ...headers,
      },
      auth,
    });
  }
}

export const sortTokens = (tokenA: string, tokenB: string): [string, string] =>
  [tokenA.toLowerCase(), tokenB.toLowerCase()].sort((a, b) =>
    a > b ? 1 : -1,
  ) as [string, string];
