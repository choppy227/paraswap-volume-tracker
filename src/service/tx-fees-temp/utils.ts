// From FE with love
interface SliceCallsInput<T, U> {
  inputArray: T[];
  execute: (inputSlice: T[], sliceIndex: number) => U;
  sliceLength: number;
}

export function sliceCalls<T, U>({
  inputArray,
  execute,
  sliceLength,
}: SliceCallsInput<T, U>): [U, ...U[]] {
  if (sliceLength >= inputArray.length) return [execute(inputArray, 0)];
  const results: U[] = [];

  for (
    let i = 0, sliceIndex = 0;
    i <= inputArray.length;
    i += sliceLength, ++sliceIndex
  ) {
    const inputSlice = inputArray.slice(i, i + sliceLength);
    const resultOfSlice = execute(inputSlice, sliceIndex);
    results.push(resultOfSlice);
  }

  return results as [U, ...U[]];
}