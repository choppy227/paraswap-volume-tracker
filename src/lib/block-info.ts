import {
  CHAIN_ID_MAINNET,
  CHAIN_ID_ROPSTEN,
  CHAIN_ID_BINANCE,
  CHAIN_ID_POLYGON,
  CHAIN_ID_FANTOM,
} from './constants';
import { Utils } from './utils';

const logger = global.LOGGER();

const SUBGRAPH_URL: { [network: number]: string } = {
  [CHAIN_ID_MAINNET]:
    'https://api.thegraph.com/subgraphs/name/blocklytics/ethereum-blocks',
  [CHAIN_ID_ROPSTEN]:
    'https://api.thegraph.com/subgraphs/name/blocklytics/ropsten-blocks',
  [CHAIN_ID_BINANCE]:
    'https://api.thegraph.com/subgraphs/name/polytoken/blocks-bsc',
  [CHAIN_ID_POLYGON]:
    'https://api.thegraph.com/subgraphs/name/polytoken/blocks-matic',
  [CHAIN_ID_FANTOM]:
    'https://thegraph.com/hosted-service/subgraph/publu/fantom-blocks',
  // 43114: TODO: deploy blocks in avalanche
};
const SUBGRAPH_TIMEOUT = 5000;

export class BlockInfo {
  blockInfo: { [block: number]: number } = {};

  constructor(private subgraphURL: string) {}

  static instances: { [network: number]: BlockInfo } = {};

  static getInstance(network: number): BlockInfo {
    if (!this.instances[network]) {
      this.instances[network] = new BlockInfo(SUBGRAPH_URL[network]);
    }
    return this.instances[network];
  }

  // This is just a special function to speed the block timestamp
  // fetching by updating the block info of a range of blocks
  async updateBlockInfo(fromBlock: number, toBlock: number) {
    // We can only get 100 entries from subgraph
    for (
      let lastFetchedBlock = fromBlock;
      lastFetchedBlock < toBlock;
      lastFetchedBlock += 100
    ) {
      // fetch the block info
      const query = `query ($number_gte: BigInt, $number_lt: BigInt) {
        blocks(orderBy: number, orderDirection: asc, where: {number_gte: $number_gte, number_lt: $number_lt}) {
          number
          timestamp
        }
      }`;
      const variables = {
        number_gte: lastFetchedBlock,
        number_lt: lastFetchedBlock + 100,
      };
      try {
        const {
          data: { data },
        } = await Utils._post(
          this.subgraphURL,
          { query, variables },
          SUBGRAPH_TIMEOUT,
        );
        data.blocks.map(
          (_block: { number: string; timestamp: string }) =>
            (this.blockInfo[parseInt(_block.number)] = parseInt(
              _block.timestamp,
            )),
        );
      } catch (e) {
        logger.error('updateBlockInfo', e);
      }
    }
  }

  // Get the timestamp of the block using blockInfo
  async getBlockTimeStamp(block: number): Promise<number | null> {
    if (block in this.blockInfo) return this.blockInfo[block];
    // if blockInfo is not available fetch the block info
    const query = `query ($block: BigInt) {
      blocks(first: 1, where: {number: $block}) {
        number
        timestamp
      }
    }`;
    const variables = {
      block,
    };
    try {
      const {
        data: { data },
      } = await Utils._post(
        this.subgraphURL,
        { query, variables },
        SUBGRAPH_TIMEOUT,
      );
      this.blockInfo[parseInt(data.blocks[0].number)] = parseInt(
        data.blocks[0].timestamp,
      );
      return parseInt(data.blocks[0].timestamp);
    } catch (e) {
      logger.error('getBlockTimeStamp', e);
      return null;
    }
  }

  // Get the blocknumber for the block after time
  async getBlockAfterTimeStamp(time: number): Promise<number | null> {
    // fetch the block info
    const query = `query ($time_gte: BigInt) {
      blocks(first: 1, orderBy: timestamp, orderDirection: asc, where: {timestamp_gte: $time_gte}) {
        number
      }
    }`;
    const variables = {
      time_gte: time,
    };
    try {
      const {
        data: { data },
      } = await Utils._post(
        this.subgraphURL,
        { query, variables },
        SUBGRAPH_TIMEOUT,
      );
      return parseInt(data.blocks[0].number);
    } catch (e) {
      logger.error('getBlockAfterTimeStamp', e);
      return null;
    }
  }
}
