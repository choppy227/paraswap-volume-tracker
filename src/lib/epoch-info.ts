import {
  CHAIN_ID_ROPSTEN,
  CHAIN_ID_MAINNET,
} from './constants';
import { BlockInfo } from './block-info';

export type StakingSetting = {
  CurrentPSPEpochReward: string;
  EpochDuration: number;
  BlockDelay: number;
  GenesisBlockNumber: number;
  CurrentEpoch: number;
};

export const StakingSettings: {
  [network: number]: StakingSetting;
} = {
  [CHAIN_ID_MAINNET]: {
    CurrentPSPEpochReward: '2500000000000000000000000',
    EpochDuration: 14 * 60 * 60 * 24,
    BlockDelay: 7,
    GenesisBlockNumber: 13620219, // This is start of the epoch 0
    CurrentEpoch: 0,
  },
  [CHAIN_ID_ROPSTEN]: {
    CurrentPSPEpochReward: '2500000000000000000000000',
    EpochDuration: 14 * 60 * 60 * 24,
    BlockDelay: 10,
    GenesisBlockNumber: 11348236, // This is start of the epoch 0
    CurrentEpoch: 0,
  },
};

// TODO: automate this
// The blocknumber at which the epoch reward is sent.
export const EpochDetails: {
  [network: number]: {
    [epoch: number]: {
      endBlockNumber: number;
      calcTimeStamp: number;
      reward: string;
    };
  };
} = {};

export class EpochInfo {
  private blockInfo: BlockInfo;

  constructor(protected network: number) {
    this.blockInfo = BlockInfo.getInstance(this.network);
  }

  static instances: {[network: number]: EpochInfo} = {};

  static getInstance(network: number): EpochInfo {
    if(!this.instances[network]) 
      this.instances[network] = new EpochInfo(network);
    return this.instances[network];
  }

  getCurrentEpoch(): number {
    return StakingSettings[this.network].CurrentEpoch;
  }

  getEpochStartBlock(epoch: number): number {
    return epoch < 1
      ? StakingSettings[this.network].GenesisBlockNumber
      : this.getEpochEndBlock(epoch - 1);
  }

  async getEpochStartCalcTime(epoch: number): Promise<number> {
    if (epoch === 0) {
      const epochStartTime = await this.blockInfo.getBlockTimeStamp(
        StakingSettings[this.network].GenesisBlockNumber,
      );
      if (!epochStartTime)
        throw new Error(
          `Unable to fetch the timestamp for network: ${this.network} block: ${
            StakingSettings[this.network].GenesisBlockNumber
          }`,
        );
      return epochStartTime;
    } else {
      return this.getEpochEndCalcTime(epoch - 1);
    }
  }

  getEpochEndCalcTime(epoch: number): number {
    return EpochDetails[this.network][epoch].calcTimeStamp;
  }

  getEpochEndBlock(epoch: number): number {
    return EpochDetails[this.network][epoch].endBlockNumber;
  }

  getCurrentPSPPoolReward(): string {
    return StakingSettings[this.network].CurrentPSPEpochReward;
  }

  getEpochDuration(): number {
    return StakingSettings[this.network].EpochDuration;
  }

  getPSPPoolReward(epoch: number): string {
    if (epoch == this.getCurrentEpoch())
      return this.getCurrentPSPPoolReward();
    return EpochDetails[this.network][epoch].reward;
  }
}