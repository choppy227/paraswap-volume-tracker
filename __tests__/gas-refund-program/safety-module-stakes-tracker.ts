import '../../src/lib/log4js';
import * as dotenv from 'dotenv';
dotenv.config();
import SafetyModuleStakesTracker from '../../scripts/gas-refund-program/staking/safety-module-stakes-tracker';
import { BNReplacer } from '../../src/lib/utils/helpers';

describe('SafetyModuleStakesTracker', () => {
  describe('snashot test for backward compat', () => {
    let tracker: SafetyModuleStakesTracker;
    const startBlock = 14567000;

    beforeAll(async () => {
      tracker = new SafetyModuleStakesTracker();
      tracker.setBlockBoundary(startBlock + 1, startBlock + 10000);
      await tracker.loadStakes();
    });

    test('Init state', () => {
      // hint: Use https://etherscan.io/tokencheck-tool to check all snapshoted data

      expect(JSON.stringify(tracker.initState, BNReplacer, 2)).toMatchSnapshot(
        `SafetyModuleStakesTracker::initState at block ${startBlock}`,
      );
    });

    test('Differential state', () => {
      expect(
        JSON.stringify(tracker.differentialStates, BNReplacer, 2),
      ).toMatchSnapshot(
        `SafetyModuleStakesTracker::differentialState between block ${tracker.startBlock} and ${tracker.endBlock}`,
      );
    });

    test('Staked PSP Balance at timestamp', () => {
      const pspBalanceAtTimestamp = tracker.computeStakedPSPBalance(
        '0x0e71f7a6bbae357a1cd364173ae69d3fb2e539e3',
        1649822860, // timestamp has to be within range
      );

      // hint: verify by computing data for earlier timestamp and compare staked balance with dapps like zerion or debank
      expect(pspBalanceAtTimestamp.toFixed(0)).toBe('31176610680399907896512');
    });
  });
});
