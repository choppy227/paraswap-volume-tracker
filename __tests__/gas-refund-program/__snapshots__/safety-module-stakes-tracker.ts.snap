// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafetyModuleStakesTracker snashot test for backward compat Differential state: SafetyModuleStakesTracker::differentialState between block 14567001 and 14577000 1`] = `
"{
  \\"stkPSPBptUsersBalances\\": {
    \\"0xfc44a13ea1a98166ffc0719f83b5f3ee2759c03f\\": [
      {
        \\"timestamp\\": 1649730091,
        \\"value\\": \\"406963502215428526485\\"
      },
      {
        \\"timestamp\\": 1649787212,
        \\"value\\": \\"357176620308005404854\\"
      }
    ],
    \\"0x8332c9c551e658fc172b55fb6093346f157b99b1\\": [
      {
        \\"timestamp\\": 1649751987,
        \\"value\\": \\"2124834581898362717002\\"
      }
    ],
    \\"0x0e71f7a6bbae357a1cd364173ae69d3fb2e539e3\\": [
      {
        \\"timestamp\\": 1649822860,
        \\"value\\": \\"3918079167367984070231\\"
      }
    ]
  },
  \\"bptPoolPSPBalance\\": [
    {
      \\"timestamp\\": 1649729977,
      \\"value\\": \\"2140255979051156275294\\"
    },
    {
      \\"timestamp\\": 1649729977,
      \\"value\\": \\"-31550791777238711495\\"
    },
    {
      \\"timestamp\\": 1649751773,
      \\"value\\": \\"13978738153352763721964\\"
    },
    {
      \\"timestamp\\": 1649751773,
      \\"value\\": \\"0\\"
    },
    {
      \\"timestamp\\": 1649761644,
      \\"value\\": \\"-29584704735718579350161\\"
    },
    {
      \\"timestamp\\": 1649786084,
      \\"value\\": \\"17652872987405754033664\\"
    },
    {
      \\"timestamp\\": 1649786856,
      \\"value\\": \\"0\\"
    },
    {
      \\"timestamp\\": 1649786856,
      \\"value\\": \\"-23312108275952523915\\"
    },
    {
      \\"timestamp\\": 1649821395,
      \\"value\\": \\"20220867064580000000000\\"
    },
    {
      \\"timestamp\\": 1649822758,
      \\"value\\": \\"20679046488460435886718\\"
    },
    {
      \\"timestamp\\": 1649822758,
      \\"value\\": \\"-10110433522609664080\\"
    }
  ],
  \\"bptPoolTotalSupply\\": [
    {
      \\"timestamp\\": 1649729977,
      \\"value\\": \\"406963502215428526485\\"
    },
    {
      \\"timestamp\\": 1649751773,
      \\"value\\": \\"2124834581898362717002\\"
    },
    {
      \\"timestamp\\": 1649786856,
      \\"value\\": \\"357176620308005404854\\"
    },
    {
      \\"timestamp\\": 1649822758,
      \\"value\\": \\"3918079167367984070231\\"
    }
  ]
}"
`;

exports[`SafetyModuleStakesTracker snashot test for backward compat Init state: SafetyModuleStakesTracker::initState at block 14567000 1`] = `
"{
  \\"stkPSPBptUsersBalances\\": {
    \\"0x0ddc793680ff4f5793849c8c6992be1695cbe72a\\": \\"128179093238755885821351\\",
    \\"0x7494eb2916cad8649f4f91eb1db6e20be605dad6\\": \\"61969992917057659809544\\",
    \\"0xf2078c58d6c38c893af4e40d7b09843ec3b7d26c\\": \\"46406495277813267849340\\",
    \\"0x933258bddd49beeca77f6d1889633c5429af45ea\\": \\"39997781113664459094671\\",
    \\"0x81ea04c179b82d31662994a68b2582ad9d860f88\\": \\"25455980928393316129539\\",
    \\"0xa932f59d834f5a5955d7a3b356a139d973afd978\\": \\"17230933158339458397951\\",
    \\"0xab205303361af614ebfce3a515fb73fd4b83d82b\\": \\"15724577857992934261657\\",
    \\"0x3ce06981bc523f950e3df346878216365b24b6fe\\": \\"15247991100000000000000\\",
    \\"0xb7d49adb031d6dbdf3e8e28f21c6dd3b6f231cd5\\": \\"12054284000000000000000\\",
    \\"0xb957dccaa1ccfb1eb78b495b499801d591d8a403\\": \\"12000868120132376782047\\",
    \\"0xb1c0d822e2cd6d564d71b50aa02fdee9d198943e\\": \\"10179513713919953506075\\",
    \\"0xc327f2f6c5df87673e6e12e674ba26654a25a7b5\\": \\"9980106395506385639117\\",
    \\"0x9a1fc77e51fa1dfae48d01ce304c250da3f4fbfa\\": \\"9402997030778617643200\\",
    \\"0x70adf879a489a3e90da1a79695a430c64bed798f\\": \\"7825560152001301134312\\",
    \\"0x9fd7be134d8d4ddb8fa8d093e348c899716f6b0b\\": \\"7017970869870899728064\\",
    \\"0xd5431fecbb14f20f8ee68360b586d84d589f9f77\\": \\"4967105622981005037160\\",
    \\"0x61229c4a3e31f1f65ce426acb10f699c6d177968\\": \\"4375248930640008375714\\",
    \\"0xc57658524232ce4879cba2e86f4754dfccff1ba3\\": \\"3997643351449740272775\\",
    \\"0x71535aae1b6c0c51db317b54d5eee72d1ab843c1\\": \\"3900103632385042954897\\",
    \\"0xfff6cb60f90582e1aaa444aecbd02f21d087edcc\\": \\"3674411042747610276840\\",
    \\"0x892c21385386b2f563f93c3f2b73191ba3442a7c\\": \\"3358559863270787126128\\",
    \\"0xa9293dca5b8135cb0c881023fd08f15ae0d9be83\\": \\"3182598121682004275834\\",
    \\"0x4677c06757a2059c7b750fd5b59c2e65bf1851b3\\": \\"2424792000000000000000\\",
    \\"0xd77900479ef30963f4b5aa7d752b15795192ee15\\": \\"2264308990062587964771\\",
    \\"0x11e73db029be1ad9f118d17783e4f8905a277c52\\": \\"2261657806021211292004\\",
    \\"0xfa610f8304909a90b1760cf8b50cc1fde6ea1f5c\\": \\"2231875556239906825478\\",
    \\"0xfc44a13ea1a98166ffc0719f83b5f3ee2759c03f\\": \\"2189153026661441158094\\",
    \\"0x0e71f7a6bbae357a1cd364173ae69d3fb2e539e3\\": \\"1988988351805200553541\\",
    \\"0x9ba961989dd6609ed091f512be947118c40f2291\\": \\"1557734913431923345618\\",
    \\"0x5d577c1cdaf838b264c7d977449c776ef664d654\\": \\"1529654935112537740139\\",
    \\"0x2c2f838f0dac3f4915f4a57f3cb63abcd806ee40\\": \\"1507305514834464308151\\",
    \\"0xbed58fa67bd966a40460b42f065dcccb0fd4fe4f\\": \\"1467708228313343560914\\",
    \\"0x39e758c943a8416e647793e8e3874da65264a0e3\\": \\"1450012600000000000000\\",
    \\"0x5f53001556e096972b5ad5bb37a77ce967e79b94\\": \\"1346230646178411790222\\",
    \\"0x8ccfbc13159c5916dbbf5815d899411aeb1876ef\\": \\"1325624345527896495836\\",
    \\"0x3538fb8136bd83dec94a508bd4f87fd43b3f5ca0\\": \\"1237646892496240611476\\",
    \\"0x38e441033c40c99a878b5058af7ec45ff4b50198\\": \\"1031205149588737676589\\",
    \\"0x2c9139d5ec9206bd779a71ecdb927c8cd42e9639\\": \\"969235000000000000000\\",
    \\"0x0951ff0835302929d6c0162b3d2495a85e38ec3a\\": \\"914284000000000000000\\",
    \\"0xfaa5869208885a15bcb87dcfd81d7f7c21ee106c\\": \\"801244034806047581091\\",
    \\"0x7235927459cfc05520b6a1467b4cd7ebf1b62217\\": \\"728734802480012602250\\",
    \\"0x2f214f4cdad32b533db27e484f12df334763c80b\\": \\"666266923186252224119\\",
    \\"0x91a90ee0e1b5e6ef19c9c50895aa4094332f2cc1\\": \\"600318500000000000000\\",
    \\"0xade8b1d612333dc80dc6f85ebf54a184ac587d5f\\": \\"488799326052298549149\\",
    \\"0x9a146f5292b4b083692804e3771297dfc665fb2a\\": \\"318448957751217252984\\",
    \\"0x3070f20f86fda706ac380f5060d256028a46ec29\\": \\"246996100000000000000\\",
    \\"0xfdc87c94917f2501123aebbb537466b80510bca9\\": \\"213276381162724572862\\",
    \\"0x05182e579fdfcf69e4390c3411d8fea1fb6467cf\\": \\"15500000000000000000\\",
    \\"0x3b4512e84017ec2dbc24e97006b47318807e1d3f\\": \\"9000000000000000000\\"
  },
  \\"bptPoolPSPBalance\\": \\"2556445774532826552913256\\",
  \\"bptPoolTotalSupply\\": \\"486095993118680093200539\\"
}"
`;

exports[`SafetyModuleStakesTracker snashot test for backward compat Init state: SafetyModuleStakesTracker::initState at block 14567001 1`] = `
"{
  \\"stkPSPBptUsersBalances\\": {
    \\"0x0ddc793680ff4f5793849c8c6992be1695cbe72a\\": \\"128179093238755885821351\\",
    \\"0x7494eb2916cad8649f4f91eb1db6e20be605dad6\\": \\"61969992917057659809544\\",
    \\"0xf2078c58d6c38c893af4e40d7b09843ec3b7d26c\\": \\"46406495277813267849340\\",
    \\"0x933258bddd49beeca77f6d1889633c5429af45ea\\": \\"39997781113664459094671\\",
    \\"0x81ea04c179b82d31662994a68b2582ad9d860f88\\": \\"25455980928393316129539\\",
    \\"0xa932f59d834f5a5955d7a3b356a139d973afd978\\": \\"17230933158339458397951\\",
    \\"0xab205303361af614ebfce3a515fb73fd4b83d82b\\": \\"15724577857992934261657\\",
    \\"0x3ce06981bc523f950e3df346878216365b24b6fe\\": \\"15247991100000000000000\\",
    \\"0xb7d49adb031d6dbdf3e8e28f21c6dd3b6f231cd5\\": \\"12054284000000000000000\\",
    \\"0xb957dccaa1ccfb1eb78b495b499801d591d8a403\\": \\"12000868120132376782047\\",
    \\"0xb1c0d822e2cd6d564d71b50aa02fdee9d198943e\\": \\"10179513713919953506075\\",
    \\"0xc327f2f6c5df87673e6e12e674ba26654a25a7b5\\": \\"9980106395506385639117\\",
    \\"0x9a1fc77e51fa1dfae48d01ce304c250da3f4fbfa\\": \\"9402997030778617643200\\",
    \\"0x70adf879a489a3e90da1a79695a430c64bed798f\\": \\"7825560152001301134312\\",
    \\"0x9fd7be134d8d4ddb8fa8d093e348c899716f6b0b\\": \\"7017970869870899728064\\",
    \\"0xd5431fecbb14f20f8ee68360b586d84d589f9f77\\": \\"4967105622981005037160\\",
    \\"0x61229c4a3e31f1f65ce426acb10f699c6d177968\\": \\"4375248930640008375714\\",
    \\"0xc57658524232ce4879cba2e86f4754dfccff1ba3\\": \\"3997643351449740272775\\",
    \\"0x71535aae1b6c0c51db317b54d5eee72d1ab843c1\\": \\"3900103632385042954897\\",
    \\"0xfff6cb60f90582e1aaa444aecbd02f21d087edcc\\": \\"3674411042747610276840\\",
    \\"0x892c21385386b2f563f93c3f2b73191ba3442a7c\\": \\"3358559863270787126128\\",
    \\"0xa9293dca5b8135cb0c881023fd08f15ae0d9be83\\": \\"3182598121682004275834\\",
    \\"0x4677c06757a2059c7b750fd5b59c2e65bf1851b3\\": \\"2424792000000000000000\\",
    \\"0xd77900479ef30963f4b5aa7d752b15795192ee15\\": \\"2264308990062587964771\\",
    \\"0x11e73db029be1ad9f118d17783e4f8905a277c52\\": \\"2261657806021211292004\\",
    \\"0xfa610f8304909a90b1760cf8b50cc1fde6ea1f5c\\": \\"2231875556239906825478\\",
    \\"0xfc44a13ea1a98166ffc0719f83b5f3ee2759c03f\\": \\"2189153026661441158094\\",
    \\"0x0e71f7a6bbae357a1cd364173ae69d3fb2e539e3\\": \\"1988988351805200553541\\",
    \\"0x9ba961989dd6609ed091f512be947118c40f2291\\": \\"1557734913431923345618\\",
    \\"0x5d577c1cdaf838b264c7d977449c776ef664d654\\": \\"1529654935112537740139\\",
    \\"0x2c2f838f0dac3f4915f4a57f3cb63abcd806ee40\\": \\"1507305514834464308151\\",
    \\"0xbed58fa67bd966a40460b42f065dcccb0fd4fe4f\\": \\"1467708228313343560914\\",
    \\"0x39e758c943a8416e647793e8e3874da65264a0e3\\": \\"1450012600000000000000\\",
    \\"0x5f53001556e096972b5ad5bb37a77ce967e79b94\\": \\"1346230646178411790222\\",
    \\"0x8ccfbc13159c5916dbbf5815d899411aeb1876ef\\": \\"1325624345527896495836\\",
    \\"0x3538fb8136bd83dec94a508bd4f87fd43b3f5ca0\\": \\"1237646892496240611476\\",
    \\"0x38e441033c40c99a878b5058af7ec45ff4b50198\\": \\"1031205149588737676589\\",
    \\"0x2c9139d5ec9206bd779a71ecdb927c8cd42e9639\\": \\"969235000000000000000\\",
    \\"0x0951ff0835302929d6c0162b3d2495a85e38ec3a\\": \\"914284000000000000000\\",
    \\"0xfaa5869208885a15bcb87dcfd81d7f7c21ee106c\\": \\"801244034806047581091\\",
    \\"0x7235927459cfc05520b6a1467b4cd7ebf1b62217\\": \\"728734802480012602250\\",
    \\"0x2f214f4cdad32b533db27e484f12df334763c80b\\": \\"666266923186252224119\\",
    \\"0x91a90ee0e1b5e6ef19c9c50895aa4094332f2cc1\\": \\"600318500000000000000\\",
    \\"0xade8b1d612333dc80dc6f85ebf54a184ac587d5f\\": \\"488799326052298549149\\",
    \\"0x9a146f5292b4b083692804e3771297dfc665fb2a\\": \\"318448957751217252984\\",
    \\"0x3070f20f86fda706ac380f5060d256028a46ec29\\": \\"246996100000000000000\\",
    \\"0xfdc87c94917f2501123aebbb537466b80510bca9\\": \\"213276381162724572862\\",
    \\"0x05182e579fdfcf69e4390c3411d8fea1fb6467cf\\": \\"15500000000000000000\\",
    \\"0x3b4512e84017ec2dbc24e97006b47318807e1d3f\\": \\"9000000000000000000\\"
  },
  \\"bptPoolPSPBalance\\": \\"2556445774532826552913256\\",
  \\"bptPoolTotalSupply\\": \\"486095993118680093200539\\"
}"
`;
